core/app/finders/spree/products/find.rb

```diff
@@ -99,7 +99,17 @@ def by_price(products)
       def by_taxons(products)
         return products unless taxons?
 
-        products.joins(:taxons).where(spree_taxons: { id: taxons })
+        ids = taxons.select { |t| t =~ /^\d+$/ }
+        permalinks = taxons + ids - (taxons & ids)
+
+        c = Spree::Taxon.connection
+
+        clause = ''
+        clause += Spree::Taxon.sanitize_sql_for_conditions(["#{c.quote_table_name('spree_taxons')}.#{c.quote_column_name('id')} IN (?)", ids]) if ids.any?
```

Comment:
```suggestion
        scope = Spree::Taxon
        if ids.any? && permalinks.any?
          scope = scope.where(id: ids.split(',')).or(scope.where(permalinks: permalinks.split(',')
        elsif ids.any?
           scope = scope.where(id: ids.split(','))
         elsif permalinks.any?
            scope = scope.where(permalink: permalinks.split(','))
        end
        
        
```

Reply:
We should avoid raw SQL as much as we can

Reply:
Believe me I tried to go down this path but couldn't make it work with the active_record `.or()`.

With hash in `or()`
```
     Failure/Error: products.joins(:taxons).where({ spree_taxons: { id: ids } }).or({ spree_taxons: { permalink: permalinks } })

     ArgumentError:
       You have passed Hash object to #or. Pass an ActiveRecord::Relation object instead.
```

With your suggestion (if I understood it right)
```
     Failure/Error: products.joins(:taxons).where({ spree_taxons: { id: ids } }).or(Spree::Taxon.where(permalink: permalinks))

     ArgumentError:
       Relation passed to #or must be structurally compatible. Incompatible values: [:joins]
```

Reply:
Looks like it's a known issue: https://github.com/rails/rails/issues/24055